## Second Assignment: Segmenting and Clustering Neighborhoods in Toronto

First step: Install packages and import libraries

In [1]:
!conda install -c conda-forge wikipedia --yes
import pandas as pd
import wikipedia as wp
import numpy as np

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    wikipedia: 1.4.0-py35_0 conda-forge

wikipedia-1.4. 100% |################################| Time: 0:00:00  17.56 MB/s


Second step: Retieve table from html source and format to read fisrt row as header

In [2]:
#Get the html source
html = wp.page("List of postal codes of Canada: M").html().encode("UTF-8")
df = pd.read_html(html)[0]
df.to_csv('beautifulsoup_pandas.csv',header=0,index=False)
# Transform first column into header
df.columns=['PostalCode', 'Borough', 'Neighborhood']
df.drop(df.index[0], inplace=True)
df.head()

PostalCode           Borough      Neighborhood
1        M1A      Not assigned      Not assigned
2        M2A      Not assigned      Not assigned
3        M3A        North York         Parkwoods
4        M4A        North York  Victoria Village
5        M5A  Downtown Toronto      Harbourfront

Third step: drop rows with borough not assigned

In [3]:
df=df[df.Borough != 'Not assigned']
df.head()

PostalCode           Borough      Neighborhood
3        M3A        North York         Parkwoods
4        M4A        North York  Victoria Village
5        M5A  Downtown Toronto      Harbourfront
6        M5A  Downtown Toronto       Regent Park
7        M6A        North York  Lawrence Heights

Third step: Group rows with identical postal code. First, in order to verify hoy many rows we should have in the end I used the group by, count() method by PostalCode

In [4]:
check=df.groupby('PostalCode').count()
check.shape

(103, 2)

This shows us there is a total of 103 different postal codes in the data frame
Then we group by postal code, but now we use the join function in order to mix different neighborhoods separated by a comma

In [5]:
new_df=df.groupby(['PostalCode', 'Borough'], as_index=False).agg(lambda x: ', '.join(set(x.astype(str))))
new_df.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Highland Creek, Port Union
2        M1E  Scarborough       Morningside, West Hill, Guildwood
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

Now we check that the total amount of rows is the same as the prdeicted one of 103

In [6]:
new_df.shape

(103, 3)

We can see that it is the same number of rows, so even though the order of the rows has changed the result is still correct

Fourth step: Replace not assigned values for neighborhood with value from borough

In [7]:
final_df=new_df
final_df['Neighborhood'].where(final_df.Neighborhood!='Not assigned', final_df.Borough, inplace=True)
final_df

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Highland Creek, Port Union  
2                    Morningside, West Hill, Guildwood  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7                      Oakridge, Golden Mile, Clairlea  
8      Cliffcrest, Scarborough Village West, Cliffside  
9                          Cliffside West, Birch Cliff  
10   Wexford Heights, Dorset Park, Scarborough Town...  
11                                   Maryvale, Wexford  
12                                           Agincourt  
13             Clarks Corners, Sullivan, Tam O'Shanter  
14   Steeles East, L'Amoreaux East, Milliken, Aginc...  
15                       Steeles West, L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                        Fairview, Henry Farm, Oriole  
19                                     Bayview Village  
20                            York Mills, Silver Hills  
21                             Newtonbrook, Willowdale  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                    Flemingdon Park, Don Mills South  
28     Wilson Heights, Bathurst Manor, Downsview North  
29                     York University, Northwood Park  
..  

To finish up, we check that there are no rows containing a Not assigned neighborhoood, that could have been joined with other neighborhoods when grouping by postal code

In [8]:
final_df[final_df['Neighborhood'].str.contains("assigned")]

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

And we verify once more the shaoe of the final dataframe

In [9]:
final_df.shape

(103, 3)